orm/primary_key.go

```diff
@@ -10,43 +12,117 @@ var _ Indexable = &PrimaryKeyTableBuilder{}
 // NewPrimaryKeyTableBuilder creates a builder to setup a PrimaryKeyTable object.
 func NewPrimaryKeyTableBuilder(prefixData byte, storeKey sdk.StoreKey, model PrimaryKeyed, codec IndexKeyCodec, cdc codec.Codec) *PrimaryKeyTableBuilder {
 	return &PrimaryKeyTableBuilder{
-		TableBuilder: NewTableBuilder(prefixData, storeKey, model, codec, cdc),
+		tableBuilder: newTableBuilder(prefixData, storeKey, model, codec, cdc),
 	}
 }
 
 type PrimaryKeyTableBuilder struct {
-	*TableBuilder
+	*tableBuilder
 }
 
 func (a PrimaryKeyTableBuilder) Build() PrimaryKeyTable {
-	return PrimaryKeyTable{table: a.TableBuilder.Build()}
+	return PrimaryKeyTable{table: a.tableBuilder.Build()}
 
 }
 
-// PrimaryKeyed defines an object type that is aware of it's immutable primary key.
+// PrimaryKeyed defines an object type that is aware of its immutable primary key.
 type PrimaryKeyed interface {
-	// PrimaryKey returns the immutable and serialized primary key of this object. The primary key has to be unique within
-	// it's domain so that not two with same value can exist in the same table.
+	// PrimaryKeyFields returns the fields of the object that will make up
+	// the primary key. The PrimaryKey function will encode and concatenate
+	// the fields to build the primary key.
+	//
+	// PrimaryKey parts can be []byte, string, and integer types. []byte is
+	// encoded with a length prefix, strings are null-terminated, and
+	// integers are encoded using 4 or 8 byte big endian.
+	//
+	// IMPORTANT: []byte parts are encoded with a single byte length prefix,
+	// so cannot be longer than 256 bytes.
 	//
-	// The `IndexKeyCodec` used with the `PrimaryKeyTable` may add certain constraints to the byte representation as
-	// max length = 255 in `Max255DynamicLengthIndexKeyCodec` or a fix length in `FixLengthIndexKeyCodec` for example.
-	PrimaryKey() []byte
+	// The `IndexKeyCodec` used with the `PrimaryKeyTable` may add certain
+	// constraints to the byte representation as max length = 255 in
+	// `Max255DynamicLengthIndexKeyCodec` or a fix length in
+	// `FixLengthIndexKeyCodec` for example.
+	PrimaryKeyFields() []interface{}
 	codec.ProtoMarshaler
 }
 
+// PrimaryKey returns the immutable and serialized primary key of this object.
+// The primary key has to be unique within it's domain so that not two with same
+// value can exist in the same table. This means PrimaryKeyFields() has to
+// return a unique value for each object.
+//
+// PrimaryKey parts can be []byte, string, and integer types. []byte is encoded
+// with a length prefix, strings are null-terminated, and integers are encoded
+// using 4 or 8 byte big endian.
+func PrimaryKey(obj PrimaryKeyed) []byte {
+	fields := obj.PrimaryKeyFields()
+	return buildPrimaryKey(fields...)
+}
+
+// buildPrimaryKey encodes and concatenates the PrimaryKeyFields. See PrimaryKey
+// for full documentation of the encoding.
+func buildPrimaryKey(fields ...interface{}) []byte {
```

Comment:
I would prefer that this function and all related functions return error rather than panic.

Reply:
I've updated the comment, and left function signature, but will be worth to consider not panicing.